From: Thorsten Alteholz <debian@alteholz.de>
Date: Wed, 13 Jan 2010 04:59:57 +0100
Subject: Debian bug 237630 error in string

This is a small patch to avoid an endless loop. In case of an error
textwrap() adds <ERR> at the end of the string and returns with the
string that was built up to that error.

Bug-Debian: http://bugs.debian.org/237630
Last-Update: 2021-09-14
---
 textwrap.3 | 3 +++
 textwrap.c | 8 ++++++++
 2 files changed, 11 insertions(+)

diff --git a/textwrap.3 b/textwrap.3
index 4bac77b..67f1f84 100644
--- a/textwrap.3
+++ b/textwrap.3
@@ -78,6 +78,9 @@ proper number of space code (0x20) by \fBtextwrap\fR.
 .SH RETURN VALUE
 \fBtextwrap()\fR returns the line-folded text.
 You can free(3) the given value.
+
+In case of any error while processing the string, the text <ERR> will be
+appended to the output and processing will be aborted.
 .\" ********************************************************************
 .SH EXAMPLE
 .nf
diff --git a/textwrap.c b/textwrap.c
index 5753ec9..0d40b9d 100644
--- a/textwrap.c
+++ b/textwrap.c
@@ -300,6 +300,14 @@ textwrap(const textwrap_t *prop, const char *text)
 
     now = p; /* current character */
     ml = mblen(p, MB_CUR_MAX);
+    if (ml<0) {
+       /* 
+        * stringt_addstr will take care about enough
+        * memory for out
+        */
+       stringt_addstr(out, " <ERR>");
+       return stringt_destroy_extract(out);
+    }
     w = mbwidth(p, ml);
     b = breakable(p, ml, encoding_type);
     p += ml;
